pro/src/data_access/ApiRecipeDataAccessObject.java

```diff
@@ -36,40 +48,35 @@ private HashMap<String, Object> filter(JSONObject recipe_json, int index) {
         HashMap<String, String> ingredients = new HashMap();
 
         for(int i = 1; i < 21; ++i) {
-            String key_ingre = "strIngredient" + Integer.toString(i);
-            String key_mea = "strMeasure" + Integer.toString(i);
-            if (recipe_json.isNull(key_mea) || index_recipe.getString(key_mea) == "") {
+            String key_ingre = "strIngredient" + i;
+            String key_mea = "strMeasure" + i;
+            if (!index_recipe.isNull(key_ingre)) {
+                if (!index_recipe.getString(key_ingre).isEmpty() && !index_recipe.getString(key_ingre).trim().isEmpty()){
+                    String var10000 = index_recipe.getString(key_mea);
+                    String contain = var10000 + " " + index_recipe.getString(key_ingre);
+                    ingredients.put(key_ingre, contain);
+                } else {
+                    break;
+                }
+            } else {
                 break;
             }
-
-            String var10000 = index_recipe.getString(key_mea);
-            String contain = var10000 + " " + index_recipe.getString(key_ingre);
-            ingredients.put(key_ingre, contain);
         }
-
         int likes = 0;
```

Comment:
By using the adapter design pattern, you can get the value of like through likeDAO and you can do so for comment.